In [1]:
import pandas as pd
import numpy as np

# CMAPSS column names
cols = (
    ["unit", "cycle"] +
    [f"op_{i}" for i in range(1, 4)] +
    [f"s_{i}" for i in range(1, 22)]
)

df = pd.read_csv(
    "..\\data\\raw\\train_FD001.txt",
    sep=r"\s+",
    header=None,
    names=cols
)

df.head()


,unit,cycle,op_1,op_2,op_3,s_1,s_2,s_3,s_4,s_5,...,s_12,s_13,s_14,s_15,s_16,s_17,s_18,s_19,s_20,s_21
0,1,1,-0.0007,-0.0004,100.0,518.67,641.82,1589.70,1400.60,14.62,...,521.66,2388.02,8138.62,8.4195,0.03,392,2388,100.0,39.06,23.4190
1,1,2,0.0019,-0.0003,100.0,518.67,642.15,1591.82,1403.14,14.62,...,522.28,2388.07,8131.49,8.4318,0.03,392,2388,100.0,39.00,23.4236
2,1,3,-0.0043,0.0003,100.0,518.67,642.35,1587.99,1404.20,14.62,...,522.42,2388.03,8133.23,8.4178,0.03,390,2388,100.0,38.95,23.3442
3,1,4,0.0007,0.0000,100.0,518.67,642.35,1582.79,1401.87,14.62,...,522.86,2388.08,8133.83,8.3682,0.03,392,2388,100.0,38.88,23.3739
4,1,5,-0.0019,-0.0002,100.0,518.67,642.37,1582.85,1406.22,14.62,...,522.19,2388.04,8133.80,8.4294,0.03,393,2388,100.0,38.90,23.4044


In [2]:
# max cycle per engine
max_cycle = df.groupby('unit')['cycle'].max().reset_index()
max_cycle.columns = ['unit', 'max_cycle']
df = df.merge(max_cycle,on='unit')
df['RUL'] = df['max_cycle'] -  df['cycle'] 
df[['unit','cycle','RUL']].head()


,unit,cycle,RUL
0,1,1,191
1,1,2,190
2,1,3,189
3,1,4,188
4,1,5,187


In [25]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import root_mean_squared_error

features = [c for c in df.columns if c.startswith("s_")]

X = df[features]
y = df["RUL"]

X_train, X_val, y_train, y_val = train_test_split(
    X, y, test_size=0.2, random_state=42
)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)

model = LinearRegression()
model.fit(X_train, y_train)

preds = model.predict(X_val)

mae = mean_absolute_error(y_val, preds)
rmse = root_mean_squared_error(y_val, preds)
print("MAE:", mae)
print(f"RMSE : {rmse}")


MAE: 34.03536855760963
RMSE : 44.32276818709505


In [4]:
features = [c for c in df.columns if c.startswith("s_")]

X = df[features]
y = df["RUL"]

X_train, X_val, y_train, y_val = train_test_split(
    X, y, test_size=0.2, random_state=42
)
y_train = y_train.clip(upper=140)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)

model = LinearRegression()
model.fit(X_train, y_train)

preds = model.predict(X_val)

mae = mean_absolute_error(y_val, preds)
rmse = root_mean_squared_error(y_val, preds)
print("MAE:", mae)
print(f"RMSE : {rmse}")


MAE: 34.77509110863931
RMSE : 48.40923538918995


In [27]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, root_mean_squared_error

# 1️⃣ Cap RUL BEFORE splitting (important)
df["RUL"] = df["RUL"].clip(upper=140)

# 2️⃣ Split by engine id
engine_ids = df["unit"].unique()

train_ids, val_ids = train_test_split(
    engine_ids, test_size=0.2, random_state=42
)

train_df = df[df["unit"].isin(train_ids)]
val_df   = df[df["unit"].isin(val_ids)]

# 3️⃣ Define features
features = [c for c in df.columns if c.startswith("s_")]

X_train = train_df[features]
y_train = train_df["RUL"]

X_val = val_df[features]
y_val = val_df["RUL"]

# 4️⃣ Scale (fit only on training)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)

# 5️⃣ Train model
model = LinearRegression()
model.fit(X_train, y_train)

# 6️⃣ Predict
preds = model.predict(X_val)

mae = mean_absolute_error(y_val, preds)
rmse = root_mean_squared_error(y_val, preds)

print("MAE:", mae)
print("RMSE:", rmse)


MAE: 18.824860250167937
RMSE: 23.492424441771934


In [29]:
# take last cycle of each validation engine
val_last = val_df.groupby("unit").last()

X_val_last = val_last[features]
y_val_last = val_last["RUL"]

X_val_last = scaler.transform(X_val_last)

preds_last = model.predict(X_val_last)

mae_last = mean_absolute_error(y_val_last, preds_last)
rmse_last = root_mean_squared_error(y_val_last, preds_last)

print("Final-cycle MAE:", mae_last)
print("Final-cycle RMSE:", rmse_last)


Final-cycle MAE: 19.085322490407474
Final-cycle RMSE: 25.5913632368481


- we can see after capping rul value at 140 the error increases.

In [24]:
from sklearn.tree import DecisionTreeRegressor

features = [c for c in df.columns if c.startswith("s_")]

X = df[features]
y = df["RUL"]

X_train, X_val, y_train, y_val = train_test_split(
    X, y, test_size=0.2, random_state=42
)
y_train = y_train.clip(upper=140)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)

model = DecisionTreeRegressor(max_depth=10)
model.fit(X_train, y_train)

preds = model.predict(X_val)

mae = mean_absolute_error(y_val, preds)
rmse = root_mean_squared_error(y_val, preds)
print("MAE:", mae)
print(f"RMSE : {rmse}")


MAE: 33.45060163544612
RMSE : 48.037342001249655


rmse and mae at different maxdepths
- md 3 ; MAE: 37.17699537400008   RMSE : 51.35069923681389
- md 4 ; MAE: 35.65842360626754   RMSE : 49.94510542120869
- md 5 ; MAE: 34.54071352840441   RMSE : 48.80853708314268
- md 6 ; MAE: 33.966775059017095  RMSE : 48.31026306873614
- md 7 ; MAE: 33.759746601819565  RMSE : 48.179578029592285
- md 8 ; MAE: 33.679925299022244  RMSE : 48.090747192010376
- md 9 ; MAE: 33.70252021731415   RMSE : 48.22454236429071
- md 10 ; MAE: 33.45060163544612  RMSE : 48.037342001249655

In [34]:
# 1️⃣ Cap RUL BEFORE splitting (important)
df["RUL"] = df["RUL"].clip(upper=140)

# 2️⃣ Split by engine id
engine_ids = df["unit"].unique()

train_ids, val_ids = train_test_split(
    engine_ids, test_size=0.2, random_state=42
)

train_df = df[df["unit"].isin(train_ids)]
val_df   = df[df["unit"].isin(val_ids)]

# 3️⃣ Define features
features = [c for c in df.columns if c.startswith("s_")]

X_train = train_df[features]
y_train = train_df["RUL"]

X_val = val_df[features]
y_val = val_df["RUL"]

# 4️⃣ Scale (fit only on training)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)

# 5️⃣ Train model
model = DecisionTreeRegressor(max_depth=6, min_samples_leaf=20, random_state=42)

model.fit(X_train, y_train)

# 6️⃣ Predict
preds = model.predict(X_val)

mae = mean_absolute_error(y_val, preds)
rmse = root_mean_squared_error(y_val, preds)

print("MAE:", mae)
print("RMSE:", rmse)


MAE: 17.414994801354712
RMSE: 23.028205824631765


In [35]:
# take last cycle of each validation engine
val_last = val_df.groupby("unit").last()

X_val_last = val_last[features]
y_val_last = val_last["RUL"]

X_val_last = scaler.transform(X_val_last)

preds_last = model.predict(X_val_last)

mae_last = mean_absolute_error(y_val_last, preds_last)
rmse_last = root_mean_squared_error(y_val_last, preds_last)

print("Final-cycle MAE:", mae_last)
print("Final-cycle RMSE:", rmse_last)


Final-cycle MAE: 6.049753516164552
Final-cycle RMSE: 7.1137945540209335
